### ISSUE TITLE
Interpretation of vulkan extensions incorrect when using --iree-vulkan-target-env flag

### ISSUE THREAD

**Issue by PhaneeshB** [iree-org/iree#11532]
### What happened?

On using the `--iree-vulkan-target-env` flag with `#vk.target_env<>` attribute assembly we see that the extensions are interpret incorrectly, ie, Extension attr is interpret as simple number causing a type mismatch assertion failure with the following [crash stack](https://gist.github.com/2e17f269c1b6385b8b3f30192470e9e0): 

```
iree-compile: iree/third_party/llvm-project/llvm/include/llvm/Support/Casting.h:566: decltype(auto) llvm::cast(const From&) [with To = mlir::iree_compiler::IREE::Vulkan::ExtensionAttr; From = mlir::Attribute]: Assertion `isa<To>(Val) && "cast<Ty>() argument of incompatible type!"' failed.

Stack dump:
0.      Program arguments: ./iree-compile --iree-input-type=none --iree-vm-bytecode-module-output-format=flatbuffer-binary --iree-hal-target-backends=vulkan --mlir-print-debuginfo "--iree-vulkan-target-env=#vk.target_env<v1.3, r(120), [VK_KHR_16bit_storage, VK_KHR_8bit_storage, VK_KHR_shader_float16_int8, VK_KHR_spirv_1_4, VK_KHR_storage_buffer_storage_class, VK_KHR_variable_pointers, VK_EXT_subgroup_size_control, VK_NV_cooperative_matrix], NVIDIA:DiscreteGPU, #vk.caps<maxComputeSharedMemorySize = 49152, maxComputeWorkGroupInvocations = 1024, maxComputeWorkGroupSize = dense<1024> : vector<3xi32>, shaderFloat64 = unit, shaderInt16 = unit, shaderInt64 = unit, subgroupFeatures = 255 : i32, subgroupSize = 32, minSubgroupSize = 32, maxSubgroupSize = 32, storageBuffer16BitAccess = unit, storagePushConstant16 = unit, uniformAndStorageBuffer16BitAccess = unit, storageBuffer8BitAccess = unit, storagePushConstant8 = unit, uniformAndStorageBuffer8BitAccess = unit, shaderFloat16 = unit, shaderInt8 = unit, variablePointersStorageBuffer = unit, variablePointers = unit, cooperativeMatrixPropertiesNV = [{aType = i8, bType = i8, cType = i32, kSize = 32 : i32, mSize = 8 : i32, nSize = 8 : i32, resultType = i32, scope = 3 : i32}, {aType = f16, bType = f16, cType = f16, kSize = 16 : i32, mSize = 16 : i32, nSize = 16 : i32, resultType = f16, scope = 3 : i32}, {aType = f16, bType = f16, cType = f32, kSize = 16 : i32, mSize = 16 : i32, nSize = 16 : i32, resultType = f32, scope = 3 : i32}]>>" --iree-stream-resource-index-bits=64 --iree-vm-target-index-bits=64 --iree-util-zero-fill-elided-attrs /home/phaneesh/IREE/iree/samples/simple_embedding/simple_embedding_test.mlir -o tmp.vmfb
...
#9 0x00007f9034f6e926 decltype(auto) llvm::cast<mlir::iree_compiler::IREE::Vulkan::ExtensionAttr, mlir::Attribute>(mlir::Attribute const&) /home/phaneesh/IREE/iree/third_party/llvm-project/llvm/include/llvm/Support/Casting.h:566:3
#10 0x00007f9034f6e952 mlir::iree_compiler::IREE::Vulkan::ExtensionAttr mlir::Attribute::cast<mlir::iree_compiler::IREE::Vulkan::ExtensionAttr>() const /home/phaneesh/IREE/iree/third_party/llvm-project/mlir/include/mlir/IR/Attributes.h:139:1
#11 0x00007f9034f68ef5 mlir::iree_compiler::IREE::Vulkan::TargetEnvAttr::ext_iterator::ext_iterator(mlir::Attribute const*)::'lambda'(mlir::Attribute)::operator()(mlir::Attribute) const /home/phaneesh/IREE/iree/compiler/src/iree/compiler/Dialect/Vulkan/IR/VulkanAttributes.cpp:118:45
#12 0x00007f9034f68f3c mlir::iree_compiler::IREE::Vulkan::TargetEnvAttr::ext_iterator::ext_iterator(mlir::Attribute const*)::'lambda'(mlir::Attribute)::_FUN(mlir::Attribute) /home/phaneesh/IREE/iree/compiler/src/iree/compiler/Dialect/Vulkan/IR/VulkanAttributes.cpp:119:11
#13 0x00007f9034f5e149 decltype(auto) llvm::callable_detail::Callable<Extension (*)(mlir::Attribute), true>::operator()<mlir::Attribute const&, 0>(mlir::Attribute const&) const /home/phaneesh/IREE/iree/third_party/llvm-project/llvm/include/llvm/ADT/STLExtras.h:317:3
#14 0x00007f9034f5e17a llvm::mapped_iterator<mlir::Attribute const*, Extension (*)(mlir::Attribute), Extension>::operator*() const /home/phaneesh/IREE/iree/third_party/llvm-project/llvm/include/llvm/ADT/STLExtras.h:421:55
#15 0x00007f9034f5f86a bool __gnu_cxx::__ops::_Iter_equals_val<Extension const>::operator()<mlir::iree_compiler::IREE::Vulkan::TargetEnvAttr::ext_iterator>(mlir::iree_compiler::IREE::Vulkan::TargetEnvAttr::ext_iterator) /usr/include/c++/9/bits/predefined_ops.h:241:20
#16 0x00007f9034f5f3b8 mlir::iree_compiler::IREE::Vulkan::TargetEnvAttr::ext_iterator std::__find_if<mlir::iree_compiler::IREE::Vulkan::TargetEnvAttr::ext_iterator, __gnu_cxx::__ops::_Iter_equals_val<Extension const> >(mlir::iree_compiler::IREE::Vulkan::TargetEnvAttr::ext_iterator, mlir::iree_compiler::IREE::Vulkan::TargetEnvAttr::ext_iterator, __gnu_cxx::__ops::_Iter_equals_val<Extension const>, std::random_access_iterator_tag) /usr/include/c++/9/bits/stl_algo.h:120:4
#17 0x00007f9034f5ee1f mlir::iree_compiler::IREE::Vulkan::TargetEnvAttr::ext_iterator std::__find_if<mlir::iree_compiler::IREE::Vulkan::TargetEnvAttr::ext_iterator, __gnu_cxx::__ops::_Iter_equals_val<Extension const> >(mlir::iree_compiler::IREE::Vulkan::TargetEnvAttr::ext_iterator, mlir::iree_compiler::IREE::Vulkan::TargetEnvAttr::ext_iterator, __gnu_cxx::__ops::_Iter_equals_val<Extension const>) /usr/include/c++/9/bits/stl_algo.h:163:5
#18 0x00007f9034f5e89c mlir::iree_compiler::IREE::Vulkan::TargetEnvAttr::ext_iterator std::find<mlir::iree_compiler::IREE::Vulkan::TargetEnvAttr::ext_iterator, Extension>(mlir::iree_compiler::IREE::Vulkan::TargetEnvAttr::ext_iterator, mlir::iree_compiler::IREE::Vulkan::TargetEnvAttr::ext_iterator, Extension const&) /usr/include/c++/9/bits/stl_algo.h:3946:5
#19 0x00007f9034f5e05e bool llvm::is_contained<llvm::iterator_range<mlir::iree_compiler::IREE::Vulkan::TargetEnvAttr::ext_iterator>, Extension>(llvm::iterator_range<mlir::iree_compiler::IREE::Vulkan::TargetEnvAttr::ext_iterator>&&, Extension const&) /home/phaneesh/IREE/iree/third_party/llvm-project/llvm/include/llvm/ADT/STLExtras.h:1869:19
#20 0x00007f9034f5d1fb mlir::iree_compiler::IREE::Vulkan::(anonymous namespace)::convertVersion(mlir::iree_compiler::IREE::Vulkan::TargetEnvAttr) /home/phaneesh/IREE/iree/compiler/src/iree/compiler/Dialect/Vulkan/Utils/TargetEnvironment.cpp:25:44
#21 0x00007f9034f5ddbc mlir::iree_compiler::IREE::Vulkan::convertTargetEnv(mlir::iree_compiler::IREE::Vulkan::TargetEnvAttr) /home/phaneesh/IREE/iree/compiler/src/iree/compiler/Dialect/Vulkan/Utils/TargetEnvironment.cpp:188:35
#22 0x00007f9033444c21 mlir::iree_compiler::IREE::HAL::getSPIRVTargetEnv(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, mlir::MLIRContext*) /home/phaneesh/IREE/iree/compiler/src/iree/compiler/Dialect/HAL/Target/VulkanSPIRV/VulkanSPIRVTarget.cpp:71:44
...
```

observations:

1. when using `--iree-vulkan-target-triple` flag it is seen that the interpretation of the extension is 
`#vk<Extension VK_KHR_16bit_storage>` 
but when using target-env flag it is interpret as `0 : i32` which is the enum value of `VK_KHR_16bit_storage` (implying incorrect interpretation of the value therefore causing a type mismatch)

2. the target-env string is parsed using `iree-opt --allow-unregistered-dialect --split-input-file --verify-diagnostics` as mentioned in [.../Vulkan/IR/test/target_env.mlir](https://github.com/iree-org/iree/blob/main/compiler/src/iree/compiler/Dialect/Vulkan/IR/test/target_env.mlir#L3)
and gives no errors (and the same assembly as output)

3. the flow for target-triple and target-env path from `convertTargetEnv(...)` onwards is exactly the same till the point of error.

4. Issue is reproduced on both Linux and MacOS

target-env attr assembly used for linux (+ Nvidia a100 GPU):

```
--iree-vulkan-target-env="#vk.target_env<v1.3, r(120), [VK_KHR_16bit_storage, VK_KHR_8bit_storage, VK_KHR_shader_float16_int8, VK_KHR_spirv_1_4, VK_KHR_storage_buffer_storage_class, VK_KHR_variable_pointers, VK_EXT_subgroup_size_control, VK_NV_cooperative_matrix], NVIDIA:DiscreteGPU, #vk.caps<maxComputeSharedMemorySize = 49152, maxComputeWorkGroupInvocations = 1024, maxComputeWorkGroupSize = dense<1024> : vector<3xi32>, shaderFloat64 = unit, shaderInt16 = unit, shaderInt64 = unit, subgroupFeatures = 255 : i32, subgroupSize = 32, minSubgroupSize = 32, maxSubgroupSize = 32, storageBuffer16BitAccess = unit, storagePushConstant16 = unit, uniformAndStorageBuffer16BitAccess = unit, storageBuffer8BitAccess = unit, storagePushConstant8 = unit, uniformAndStorageBuffer8BitAccess = unit, shaderFloat16 = unit, shaderInt8 = unit, variablePointersStorageBuffer = unit, variablePointers = unit, cooperativeMatrixPropertiesNV = [{aType = i8, bType = i8, cType = i32, kSize = 32 : i32, mSize = 8 : i32, nSize = 8 : i32, resultType = i32, scope = 3 : i32}, {aType = f16, bType = f16, cType = f16, kSize = 16 : i32, mSize = 16 : i32, nSize = 16 : i32, resultType = f16, scope = 3 : i32}, {aType = f16, bType = f16, cType = f32, kSize = 16 : i32, mSize = 16 : i32, nSize = 16 : i32, resultType = f32, scope = 3 : i32}]>>"
```

### Steps to reproduce your issue

For Linux with Nvidia A100 GPU

```
../iree-build/tools/iree-compile \
--iree-input-type=none \
--iree-vm-bytecode-module-output-format=flatbuffer-binary \
--iree-hal-target-backends=vulkan \
--mlir-print-debuginfo \
--iree-vulkan-target-env="#vk.target_env<v1.3, r(120), [VK_KHR_16bit_storage, VK_KHR_8bit_storage, VK_KHR_shader_float16_int8, VK_KHR_spirv_1_4, VK_KHR_storage_buffer_storage_class, VK_KHR_variable_pointers, VK_EXT_subgroup_size_control, VK_NV_cooperative_matrix], NVIDIA:DiscreteGPU, #vk.caps<maxComputeSharedMemorySize = 49152, maxComputeWorkGroupInvocations = 1024, maxComputeWorkGroupSize = dense<1024> : vector<3xi32>, shaderFloat64 = unit, shaderInt16 = unit, shaderInt64 = unit, subgroupFeatures = 255 : i32, subgroupSize = 32, minSubgroupSize = 32, maxSubgroupSize = 32, storageBuffer16BitAccess = unit, storagePushConstant16 = unit, uniformAndStorageBuffer16BitAccess = unit, storageBuffer8BitAccess = unit, storagePushConstant8 = unit, uniformAndStorageBuffer8BitAccess = unit, shaderFloat16 = unit, shaderInt8 = unit, variablePointersStorageBuffer = unit, variablePointers = unit, cooperativeMatrixPropertiesNV = [{aType = i8, bType = i8, cType = i32, kSize = 32 : i32, mSize = 8 : i32, nSize = 8 : i32, resultType = i32, scope = 3 : i32}, {aType = f16, bType = f16, cType = f16, kSize = 16 : i32, mSize = 16 : i32, nSize = 16 : i32, resultType = f16, scope = 3 : i32}, {aType = f16, bType = f16, cType = f32, kSize = 16 : i32, mSize = 16 : i32, nSize = 16 : i32, resultType = f32, scope = 3 : i32}]>>" \
--iree-stream-resource-index-bits=64 \
--iree-vm-target-index-bits=64 \
--iree-util-zero-fill-elided-attrs ./samples/simple_embedding/simple_embedding_test.mlir -o tmp.vmfb
```

For MacOS (M1 / M2)
```
../iree-build/tools/iree-compile \
--iree-input-type=none \
--iree-vm-bytecode-module-output-format=flatbuffer-binary \
--iree-hal-target-backends=vulkan \
--mlir-print-debuginfo \
--mlir-print-op-on-diagnostic=false \
--iree-llvm-target-cpu-features=host \
--iree-vulkan-target-env="#vk.target_env<v1.3, r(120), [VK_KHR_16bit_storage, VK_KHR_8bit_storage, VK_KHR_shader_float16_int8, VK_KHR_storage_buffer_storage_class, VK_KHR_variable_pointers], Apple:IntegratedGPU, #vk.caps<maxComputeSharedMemorySize = 32768, maxComputeWorkGroupInvocations = 1024, maxComputeWorkGroupSize = dense<1024> : vector<3xi32>, shaderFloat64 = unit, shaderInt16 = unit, shaderInt64 = unit, subgroupFeatures = 191 : i32, subgroupSize = 32, storageBuffer16BitAccess = unit, storagePushConstant16 = unit, uniformAndStorageBuffer16BitAccess = unit, storageBuffer8BitAccess = unit, storagePushConstant8 = unit, uniformAndStorageBuffer8BitAccess = unit, shaderFloat16 = unit, shaderInt8 = unit, variablePointersStorageBuffer = unit, variablePointers = unit>>"  \
--iree-stream-resource-index-bits=64 \
--iree-vm-target-index-bits=64 \
--iree-util-zero-fill-elided-attrs \ 
./samples/simple_embedding/simple_embedding_test.mlir -o simpletmp.vmfb
```

**Comment by powderluv** [comment]
@antiagainst  fyi

**Comment by antiagainst** [comment]
Yeah the path is broken. I will come up with a fix later.

**Comment by antiagainst** [comment]
This should be fixed by #11567. Note that the format changed a bit; please see the test update in the pull request as examples.
